Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp

```diff
@@ -314,51 +361,81 @@ std::optional<Vector<LayoutUnit>> InlineContentBalancer::balanceRangeWithNoLineR
     // Compute final line widths
     Vector<LayoutUnit> lineWidths(breaks.size());
     for (size_t i = 0; i < breaks.size(); i++) {
-        auto start = i > 0 ? breaks[i - 1] : range.startIndex();
+        auto start = !i ? range.startIndex() : breaks[i - 1];
         auto end = breaks[i];
-        SlidingWidth slidingWidth { *this, m_inlineItems, start, end };
-        lineWidths[i] = LayoutUnit::fromFloatCeil(slidingWidth.width() + LayoutUnit::epsilon());
+        auto indentWidth = !i ? firstLineTextIndent : textIndent;
+        SlidingWidth slidingWidth { *this, m_inlineItems, start, end, !i && isFirstChunk, !i };
+        lineWidths[i] = LayoutUnit::fromFloatCeil(indentWidth + slidingWidth.width() + LayoutUnit::epsilon());
     }
 
     return lineWidths;
 }
 
-InlineLayoutUnit InlineContentBalancer::inlineItemWidth(size_t inlineItemIndex) const
+InlineLayoutUnit InlineContentBalancer::inlineItemWidth(size_t inlineItemIndex, bool useFirstLineStyle) const
 {
-    return m_inlineItemWidths[inlineItemIndex];
+    return useFirstLineStyle ? m_firstLineStyleInlineItemWidths[inlineItemIndex] : m_inlineItemWidths[inlineItemIndex];
 }
 
-bool InlineContentBalancer::shouldTrimLeading(size_t inlineItemIndex) const
+bool InlineContentBalancer::shouldTrimLeading(size_t inlineItemIndex, bool useFirstLineStyle, bool isFirstLineInChunk) const
 {
-    const auto& inlineItem = m_inlineItems[inlineItemIndex];
-    auto itemWidth = inlineItemWidth(inlineItemIndex);
+    auto& inlineItem = m_inlineItems[inlineItemIndex];
+    auto& style = useFirstLineStyle ? inlineItem.firstLineStyle() : inlineItem.style();
 
-    if (inlineItem.isText() && downcast<InlineTextItem>(inlineItem).isWhitespace())
+    // Handle line break first so we can focus on other types of white space
+    if (inlineItem.isLineBreak())
         return true;
-    if (itemWidth <= 0)
+
+    if (inlineItem.isText()) {
+        auto& textItem = downcast<InlineTextItem>(inlineItem);
+        if (textItem.isWhitespace()) {
+            bool isFirstLineLeadingPreservedWhiteSpace = style.whiteSpaceCollapse() == WhiteSpaceCollapse::Preserve && isFirstLineInChunk;
+            if (isFirstLineLeadingPreservedWhiteSpace || style.whiteSpaceCollapse() == WhiteSpaceCollapse::BreakSpaces)
+                return false;
+            return true;
```

Comment:
I'd collapse this to
return !isFirstLineLeadingPreservedWhiteSpace && style.whiteSpaceCollapse() ! WhiteSpaceCollapse::BreakSpaces;